### PR TITLE
Executor 2.0: Graph lowering

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
Graph lowering takes an abstract graph definition (`graph::OpGraph`) and converts it to the execution-level "lowered" graph. The lowering is done by traversing the "definition" graph and replicating its structure.

This PR also introduces more complex tests, where a complete pipeline is defined as an `graph::OpGraph`, lowered and executed.

### Affected modules and functionalities:
None; new functionality.

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4030

### CODE REVIEW

**Review by mdabek-nvidia — APPROVED** [review]

File: dali/pipeline/executor/executor2/exec2_test.h [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+  b.Add("op1c", std::move(AddCommonArgs(spec1c, 32, "mixed", 1)));
+  b.Add("op2",  std::move(AddCommonArgs(spec2,  32, "cpu", 1)));
+  b.Add("op2c", std::move(AddCommonArgs(spec2c, 32, "mixed", 1)));
+  b.Add("op3",  std::move(AddCommonArgs(spec3,  32, "gpu", 1)));
+  b.AddOutput("op3_0_gpu");
+  b.AddOutput("op2_0_gpu");
+  return std::move(b).GetGraph(true);
+}
+
+inline void CheckTestGraph2Results(const Workspace &ws, int batch_size) {
```
Comment (mdabek-nvidia) [inline dali/pipeline/executor/executor2/exec2_test.h:228]: I would add graph that executes GPU->CPU flow. May be in one of the next PRs.

Comment (mzient) [reply dali/pipeline/executor/executor2/exec2_test.h:134]: I'll add those when we have a GPU->CPU operator.

File: dali/pipeline/executor/executor2/exec_graph_test.cc [hunk 2]
```diff
@@ -397,6 +397,86 @@ TEST(ExecGraphTest, Exception) {
...
+TEST(ExecGraphTest, LoweredStructureMatch) {
+  graph::OpGraph def = GetTestGraph1();
+  ExecGraph g;
+  g.Lower(def);
+  ASSERT_EQ(g.Nodes().size(), def.OpNodes().size() + 1);
+  EXPECT_TRUE(g.Nodes().back().is_pipeline_output);
+  EXPECT_EQ(g.Nodes().back().inputs.size(), 2_uz);
+  auto def_it = def.OpNodes().begin();
+  auto ex_it = g.Nodes().begin();
+  for (; def_it != def.OpNodes().end(); def_it++, ex_it++) {
```
Comment (szkarpinski) [inline dali/pipeline/executor/executor2/exec_graph_test.cc:409]: Nitpick: skipping the first expression in `for` and using a coma in the last one makes this line quite hard to read :/

**Review by szkarpinski — APPROVED** [review]

File: dali/pipeline/executor/executor2/exec2_test.h [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+inline void CheckTestGraph1Results(const Workspace &ws, int batch_size) {
+  auto &o0 = ws.Output<CPUBackend>(0);
+  auto &o1 = ws.Output<CPUBackend>(1);
+  ASSERT_EQ(o0.num_samples(), batch_size);
+  ASSERT_EQ(o1.num_samples(), batch_size);
+  for (int i = 0; i < batch_size; i++) {
+    // The pipeline:
+    // op0 = DummyOp(addend=10)
```
Comment (szkarpinski) [inline dali/pipeline/executor/executor2/exec2_test.h:76]: Is this comment up to date? There's no o2, o3

**Pull request merged**